aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h

```diff
@@ -0,0 +1,388 @@
+
+    // NOTE: unavoidable sync to get to know the result's shape.
+    const auto intersection_nnz = static_cast<int64_t>(
+        // shifted_offset is a 1-dim tensor, potentially empty
+        shifted_offset.size(0)
+        ? shifted_offset.select(-1, -1).template item<offset_t>()
+        : 0);
+
+    auto selected_buffer = at::empty({2, intersection_nnz}, intersection_count.options());
+    auto selected_source = selected_buffer.select(0, 0);
+    auto selected_probably_coalesced = selected_buffer.select(0, 1);
+    const auto source_idx = nnz_arange.narrow(-1, 0, source._nnz());
+    auto dummy = at::empty({1}, source_idx.options());
+
+    auto iter = TensorIteratorConfig()
+      .set_check_mem_overlap(false)
+      .check_all_same_dtype(false)
+      .add_owned_output(dummy.expand_as(source_idx))
+      .add_input(source_idx) // index_t
+      .add_input(intersection_count) // hash_t
+      .add_input(intersection_first_idx) // hash_t
+      .add_input(shifted_offset) // offset_t
+      .build();
+
+    AT_DISPATCH_INDEX_TYPES(source_idx.scalar_type(), NAME, [&]() {
+        auto* RESTRICT ptr_selected_source = selected_source.data_ptr<hash_t>();
+        auto* RESTRICT ptr_selected_probably_coalesced = selected_probably_coalesced.data_ptr<hash_t>();
+        const auto* RESTRICT ptr_argsort = argsort_hash.data_ptr<index_t>();
+        KernelLauncher::launch(iter, [=] FUNCAPI (
+              index_t idx,
+              hash_t count,
+              hash_t first_match_idx,
+              offset_t shifted_offset) -> index_t {
+            const auto offset = shifted_offset - static_cast<offset_t>(count);
+            auto* RESTRICT ptr_selected_source_idx_out = ptr_selected_source + offset;
+            auto* RESTRICT ptr_selected_probably_coalesced_idx_out = ptr_selected_probably_coalesced + offset;
+            const auto* RESTRICT ptr_argsort_idx = ptr_argsort + first_match_idx;
+            for (hash_t i = 0; i < count; ++i) {
+              *ptr_selected_source_idx_out++ = idx;
+              *ptr_selected_probably_coalesced_idx_out++ = *ptr_argsort_idx++;
+            }
+
+            return 0;
+        });
+    });
+
+    return std::make_tuple(selected_source, selected_probably_coalesced);
+  }();
+
+  const auto res_indices = source._indices().index_select(1, selected_source);
+  const auto selected_source_values = source._values().index_select(0, selected_source);
+  const auto selected_probably_coalesced_values = probably_coalesced._values().index_select(0, selected_probably_coalesced);
+  const auto res_values = binary_op_t::apply(selected_source_values, selected_probably_coalesced_values);
+  const auto res_sparse_dim = source.sparse_dim();
+  const auto res_dense_dim = res_values.dim() - 1;
+  const auto res_shape = broadcasted_shape;
+  const auto res_nnz = selected_source_values.size(0);
+
+  get_sparse_impl(res)->raw_resize_(res_sparse_dim, res_dense_dim, res_shape);
```

Comment:
I think we talked about this before and @bhosmer pointed out that `get_sparse_impl` won't be inlined. Should we store/reuse the impl ptr in this section? 

This is a common pattern, but I think that we may want to stop repeating it if the function call is non-trival.

Reply:
Agree, a pointer sounds good to me.